### PR TITLE
Onoff action to change Hound weapons and high/low trajectory

### PR DESCRIPTION
**Pull request by L-e-x-o-n**
Made as requested on Discord #keybinds channel, this widget makes it possible to use the default 'onoff' action to change between Hound weapons and artillery low/high trajectories.

'onoff' is used for activating/deactivating units like solars and defences like Bulwark and Pulsar.
Armada Hound and popup artilleries with low and high trajectory mode also use on/off to switch between states (https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/luarules/gadgets/cmd_onoffdesc.lua), but the existing 'onoff' action couldn't be bound to a key and used to switch between weapons or trajectories. 

This widget extends the 'onoff' action by making it possible to toggle weapons and trajectories without changing it's current behaviour. The default keybind for 'onoff' is X which will now also be able to switch weapons/trajectories.

### CODE REVIEW

**Comment by WatchTheFort** [comment]
This should be incorporated into the cmd_onoffdesc gadget, rather than added as an additional widget. Does this contain any gadget-incompatible code?

**Comment by L-e-x-o-n** [comment]
> This should be incorporated into the cmd_onoffdesc gadget, rather than added as an additional widget. Does this contain any gadget-incompatible code?

I can move it to gadget. I think there is no gadget-incompatible code.

**Comment by L-e-x-o-n** [comment]
Ok I was wrong, there is gadget incompatible code, I tried moving it to gadget but it doesn't work because of line 38 `widgetHandler:AddAction("onoff", onoff, nil, "t")`

File: luaui/Widgets/cmd_onoff.lua [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+local spGetSelectedUnits = Spring.GetSelectedUnits
+local spGetUnitStates = Spring.GetUnitStates
+local spGiveOrderToUnit = Spring.GiveOrderToUnit
+
+local function onoff()
+	local selectedUnits = spGetSelectedUnits()
+
+	if selectedUnits[1] == nil then return end
+
+	local weapon = spGetUnitStates(selectedUnits[1])["active"]
```
Comment (WatchTheFort) [inline luaui/Widgets/cmd_onoff.lua:22]: If the unit isn't `onoffable`, will this return nil?

File: luaui/Widgets/cmd_onoff.lua [hunk 2]
```diff
@@ -0,0 +1,39 @@
+function widget:GetInfo()
+	return {
+		name = "Onoff for Hound and trajectory",
+		desc = "onoff action will now also switch between Hound weapons and high/low trajectory",
```
Comment (WatchTheFort) [inline luaui/Widgets/cmd_onoff.lua:4]: Should be a generic description, that this applies to all units affected by the cmd_onoffdesc gadget.

File: luaui/Widgets/cmd_onoff.lua [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		weapon = 0
+	elseif weapon == false then
+		weapon = 1
+	else
+		return
+	end
+
+	for _, unit in pairs(selectedUnits) do
+		spGiveOrderToUnit(unit, CMD.ONOFF, { weapon }, 0)
+	end
```
Comment (WatchTheFort) [inline luaui/Widgets/cmd_onoff.lua:34]: Does this have performance issues if a lot of units are selected? If 1,000 units are selected, but the On/Off command only applies to one, it is still trying to issue the order to the other 999.

Comment (badosu) [reply luaui/Widgets/cmd_onoff.lua:34]: It's possible this is more performant than checking, as it avoids invoking `GetUnitStates` for each unit, delegating the check to the engine (or some callin). Only profiling can tell

**Comment by badosu** [comment]
You might want to `return true` at the end of the method to halt the chain and signal it performed a successful action. Otherwise the next actions bound to the same key might trigger.

File: luaui/Widgets/cmd_onoff.lua (hunk 3 above)
Comment (WatchTheFort) [reply luaui/Widgets/cmd_onoff.lua:34]: Issuing an order to a unit invokes additional callins, particularly `CommandFallback` and `AllowCommand`, so it is probably better to prune beforehand

File: luaui/Widgets/cmd_onoff.lua [hunk 4]
```diff
@@ -13,26 +13,23 @@ end
...
-	else
-		return
-	end
+	local firstOnoff = nil
 
 	for _, unit in pairs(selectedUnits) do
-		spGiveOrderToUnit(unit, CMD.ONOFF, { weapon }, 0)
+		local unitDefID = spGetUnitDefID(unit)
+
+		if UnitDefs[unitDefID].onOffable == true then
```
Comment (badosu) [inline luaui/Widgets/cmd_onoff.lua:25]: `== true` is not necessary. Not a blocker nevertheless

Comment (L-e-x-o-n) [reply luaui/Widgets/cmd_onoff.lua]: > `== true` is not necessary. Not a blocker nevertheless

Fixed

**Comment by badosu** [comment]
LGTM :+1:

**Comment by badosu** [comment]
One last remark, there's also `Spring.GetSelectedUnitsSorted` which returns a table indexed by unitdefid which could save you tons of calls and checks by both avoiding n+1 from `GetUnitDefID` and batches of units inside the loop.

**Comment by L-e-x-o-n** [comment]
> One last remark, there's also [Spring.GetSelectedUnitsSorted](https://springrts.com/wiki/Lua_UnsyncedRead#GetSelectedUnitsSorted) which returns a table indexed by unitdefid that can save you tons of calls and checks by both avoiding n+1 from `GetUnitDefID` and batches of units inside the loop.
> 
> Then it's a matter of checking it respects the unit being onoffable present in ordermenu with this method as opposed to the previous.

Good idea, switched to using [Spring.GetSelectedUnitsSorted](https://springrts.com/wiki/Lua_UnsyncedRead#GetSelectedUnitsSorted) doesn't seem to return what the documentation says or I missunderstood it, did some testing and got it working.

**Comment by L-e-x-o-n** [comment]
Sorry for review request spam I didn't realise I can only do it for 1 person

**Comment by WatchTheFort** [comment]
The more I look at this, the more I realize the proper course of action is to implement high/low trajectory and Hound weapon switching as distinct commands, rather than overloading the On/Off command.

File: luaui/Widgets/cmd_onoff.lua [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+	for unitDefId, units in pairs(selectedUnitsSorted) do
+		--Actual return doesn't seem to match documentation
+		--First element is of selectedUnitsSorted is n = numberOfSelectedUnits
+		--Skip that
+		if unitDefId ~= "n" then
+			if UnitDefs[unitDefId].onOffable == true then
+				if firstOnoff == nil then
+					local isActive = spGetUnitStates(units[1])["active"]
+					if isActive then firstOnoff = 0 else firstOnoff = 1 end
```
Comment (WatchTheFort) [inline luaui/Widgets/cmd_onoff.lua:31]: The behaviour should match what happens with other stateful orders when the units selected have a mix of states. Can you test what happens with e.g. jammers when issuing the on/off command to a group of them who have a mix of on and off states?
Does it switch based on the first unit selected?

Comment (L-e-x-o-n) [reply luaui/Widgets/cmd_onoff.lua:31]: Yes, it switches all selected units based on the first unit with on off.

**Comment by badosu** [comment]
> The more I look at this, the more I realize the proper course of action is to implement high/low trajectory and Hound weapon switching as distinct commands, rather than overloading the On/Off command.

Agreed!

File: luaui/Widgets/cmd_onoff.lua [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+local function onoff()
+	--Should return { [number unitDefID] = { [1] = [number unitID], etc... }, ... }
+	local selectedUnitsSorted = spGetSelectedUnitsSorted()
+	local firstOnoff = nil
+
+	for unitDefId, units in pairs(selectedUnitsSorted) do
+		--Actual return doesn't seem to match documentation
+		--First element is of selectedUnitsSorted is n = numberOfSelectedUnits
+		--Skip that
+		if unitDefId ~= "n" then
```
Comment (badosu) [inline luaui/Widgets/cmd_onoff.lua:27]: Nice catch!

**Pull request merged**